tower-http/src/services/fs/parse_range.rs

```diff
@@ -0,0 +1,309 @@
+                    "Range: {} is not acceptable, contains multiple dashes (-).",
+                    range
+                ));
+            }
+            if let Some((start, end)) = split_once(range, "-") {
+                if start == "" {
+                    if let Ok(end) = end.parse::<u64>() {
+                        if end >= file_size_bytes {
+                            return ParsedRangeHeader::Unsatisfiable(format!(
+                                "Range: {} is not satisfiable, end of range exceeds file boundary.",
+                                range
+                            ));
+                        }
+                        if end == 0 {
+                            return ParsedRangeHeader::Unsatisfiable(format!("Range: {} is not satisfiable, suffixed number of bytes to retrieve is zero.", range));
+                        }
+                        let start = file_size_bytes - 1 - end;
+                        ranges.push(RangeInclusive::new(start, file_size_bytes - 1));
+                        continue;
+                    }
+                    return ParsedRangeHeader::Malformed(format!(
+                        "Range: {} is not acceptable, end of range not parseable.",
+                        range
+                    ));
+                }
+                if let Ok(start) = start.parse::<u64>() {
+                    if end == "" {
+                        ranges.push(RangeInclusive::new(start, file_size_bytes - 1));
+                        continue;
+                    }
+                    if let Ok(end) = end.parse::<u64>() {
+                        if end >= file_size_bytes {
+                            return ParsedRangeHeader::Unsatisfiable(format!(
+                                "Range: {} is not satisfiable, end of range exceeds file boundary.",
+                                range
+                            ));
+                        }
+                        ranges.push(RangeInclusive::new(start, end));
+                        continue;
+                    }
+                    return ParsedRangeHeader::Malformed(format!(
+                        "Range: {} is not acceptable, end of range not parseable.",
+                        range
+                    ));
+                }
+                return ParsedRangeHeader::Malformed(format!(
+                    "Range: {} is not acceptable, start of range not parseable.",
+                    range
+                ));
+            }
+            return ParsedRangeHeader::Malformed(format!(
+                "Range: {} is not acceptable, range does not contain any dashes.",
+                range
+            ));
+        }
+    } else {
+        return ParsedRangeHeader::Malformed(format!(
+            "Range: {} is not acceptable, range does not start with 'bytes='",
+            range
+        ));
+    }
+    if ranges.is_empty() {
+        panic!("Programming error parsing range {}", range)
+    } else {
```

Comment:
We should aim for `tower-http` _never_ to panic because of user input. If someone can make the server panic by sending a specific header thats technically an attack vector that can be used to do DoS.

Reply:
Yeah, this should(TM) be impossible though as a bad header should always be a Malformed or Unsatisfiable, any valid header contains a range, scratching my head over this as something needs to be returned but if it's a Malformed then that's misleading because that should have been caught earlier, Unsatisfiable as well. If we end up here it's because I wrote a bug

Reply:
Returning `Malformed` with the body saying something like "something went wrong in tower-http, please file an issue" is fine imo.

I'm mostly worried about this code getting refactored in the future such that this is no longer unreachable.